### PR TITLE
feat: make truely multi-root aware

### PR DESCRIPTION
**Pull request by nikku**
This adds the capability to add multiple root elements to the Canvas. 
As a side-effect, it drops our leaky plane abstraction from the public 
API.

The following `Canvas` APIs make the magic happen:

* `Canvas#addRootElement` - add a root element without showing it

* `Canvas#removeRootElement` - remove a root element

* `Canvas#setRootElement` - set to a specific root element; for 
  backwards compatibility the add-during-set semantic is kept

* `Canvas#getRootElements` - returns a list of root elements that can 
  be switched to

Under the hood we continue to rely on root element specific layers.
However, we clean these layers up once the root element is removed.

---

#### BREAKING CHANGES:

* All plane related APIs got removed, use the newly introduced
  `(add|set)RootElement` APIs to accomplish the same thing.
* INTERNAL: Every root element gets it's own layer now; there is no 
  magic re-use of the BASE_LAYER happening anymore.
* `setRootElement` does not have single root semantics anymore. As such, 
  it does not blow up if a non-existing root is being passed; rather, it 
  adds that new root and shows it.
* `setRootElement` has on `override` semantics anymore. To replace the 
  current root, set a new root and remove the old one.

----

#### TODO

* [x] Don't make `set***` have add semantics

---

Closes https://github.com/bpmn-io/diagram-js/issues/600

### CODE REVIEW

**Comment by nikku** [comment]
@marstamm To be examined.

**Comment by nikku** [comment]
I've refined the existing API, adding a `Canvas#getRootElements` method. It is the natural partner for the other *rootElement APIs.

File: lib/core/Canvas.js
Comment (marstamm) [inline lib/core/Canvas.js:130]: I think we should could get rid of planes (and rendering on different layers) completely. This way, all root elements are rendered on the same layer and we don't run into problems with utility planes. We can track the graphics element using the elementRegistry and don't have to worry about renaming layers when we the ID changes.

I added a proposal of how this could look like here: 5a516149b0bef0b3806cdb01d4500da27d0c2db9

Comment (marstamm) [reply lib/core/Canvas.js:130]: My initial Idea was to handle root elements like normal elements and use the `hidden` attribute to hide/show them, but changing the flag is a modeling action that would not work in the viewer

Comment (nikku) [reply lib/core/Canvas.js:130]: Yea. I did think along the same lines.

What is great about this PR, regardless of this discussion is the fact that how multi-root is implemented is _entirely_ implementation detail.

We can follow up on stuff like https://github.com/bpmn-io/diagram-js/commit/5a516149b0bef0b3806cdb01d4500da27d0c2db9 without any breaking changes.

Comment (nikku) [reply lib/core/Canvas.js:130]: Would be interested in your judgement: Is all-roots-on-the-samne-layer feasible?

Comment (nikku) [reply lib/core/Canvas.js:130]: The tests on your branch fail, albeit with a silly lint error: https://github.com/bpmn-io/diagram-js/runs/4481927647?check_suite_focus=true#step:6:17

Comment (marstamm) [reply lib/core/Canvas.js:130]: Yeah, they don't work (yet). I wanted to timebox looking into it and there were 2 tests left after it was up

I think it is feasible, my diagram-js + your bmn-js branch produce a usable diagram, even when some tests fail

Comment (marstamm) [reply lib/core/Canvas.js:130]: which also shows that the API is strong, if we can fundamentally change things in diagram-js and bpmn-js still works fine

Comment (nikku) [reply lib/core/Canvas.js:130]: So where do we go from here? I'd say we consider this PR for merging and rework the internals at a later stage?

Comment (nikku) [reply lib/core/Canvas.js:130]: As [I said](https://github.com/bpmn-io/diagram-js/pull/601#discussion_r766543345); this is the main purpose of the API.

Comment (marstamm) [reply lib/core/Canvas.js:130]: Yes, let's do it this way :+1:

File: lib/core/Canvas.js
Comment (marstamm) [inline lib/core/Canvas.js:1094]:
```suggestion
  // set to correct rootElement
```

File: lib/core/Canvas.js
Comment (marstamm) [inline lib/core/Canvas.js:629]: Why do we bind adding the element to `override`? If we support adding the element implicitly, I think we should make do it consistently when the element was not added before.

File: lib/core/Canvas.js
Comment (marstamm) [inline lib/core/Canvas.js:127]: Is there any specific reasons for this number or is it random?

**Comment by nikku** [comment]
Before we merge this one I'd appreciate if you could assert it from the modeling (BPMN perspective). There is a [piece of dead code inside](https://app.codecov.io/gh/bpmn-io/diagram-js/compare/601/changes#D2L342) that we'd likely need? Or maybe not, if updating an elements id (`element.updateId`) does automatically update the linked layer.

**Review by marstamm — COMMENTED** [review]
I really like the new API.
The only thing that feels weird to me is the `override` flag on `#setRootElement`. I'm not sure what the expected behavior in a multi-root environment is, as we set the root element either way.
In an Ideal world, I would like to get rid of it and always use `removeRootElement` explicitly. Would that be feasible? Am I missing something?

File: lib/core/Canvas.js
Comment (nikku) [reply lib/core/Canvas.js:127]: It is random, larger than `-1`.

File: lib/core/Canvas.js
Comment (nikku) [reply lib/core/Canvas.js:629]: The idea was to keep API compatibility with diagram-js < 7.4.

We can decide to ditch it all-together. Would be interesting to get your feedback.

Comment (nikku) [reply lib/core/Canvas.js:629]: The use case is to "replace an existing root, in place".

Cf. bpmn-js Process -> collaboration.

Comment (marstamm) [reply lib/core/Canvas.js:629]: I would like to ditch it. Replace in place is still possible, but needs 2 canvas operations:
```
  canvas.setRootElement(newRoot);
  canvas.removeRootElement(oldRoot);
```

Using the API this way also makes it more robust against edge cases (e.g. active plane is a subprocess and the rootProcess is replaced with a collaboration via API)

Comment (marstamm) [reply lib/core/Canvas.js:629]: The removing of the root element still has not other side effects, so the adjustments in the `UpdateCanvasRootHandler` is straight forward

Comment (nikku) [reply lib/core/Canvas.js:629]: Makes sense. I'll do that then. When is a better time to ditch then now, as we rework the API entirely?

Comment (marstamm) [reply lib/core/Canvas.js:629]: Yes, lets ditch it now. We have breaking changes anyway, so we can include it in diagram-js@8

File: lib/core/Canvas.js
Comment (marstamm) [reply lib/core/Canvas.js:127]: Why not start the count at 0 then?

Comment (nikku) [reply lib/core/Canvas.js:127]: If that causes confusion, let#s do it. 

image

Comment (barmac) [reply lib/core/Canvas.js:127]: Let's just clarify it via a comment:

```suggestion
  // random number with no significance when it comes to the value
  this._roots = 79131;
```

Comment (nikku) [reply lib/core/Canvas.js]: Fixed via https://github.com/bpmn-io/diagram-js/pull/601/commits/cd2e78221b7aed0d8980e438798ed63ff25a4da9.

File: lib/core/Canvas.js
Comment (nikku) [reply lib/core/Canvas.js]: Ditched via https://github.com/bpmn-io/diagram-js/pull/601/commits/bd2149b951d9b79fcb406b3b20ed2690046bdad3.

Also, we do throw a descriptive error message on miss-use.

File: lib/core/Canvas.js
Comment (nikku) [reply lib/core/Canvas.js]: :haircut:

**Comment by nikku** [comment]
Accounted for your review hints @marstamm.

**Review by marstamm — APPROVED** [review]
Thank you for the quick follow ups and looking into this!

Looks good to me now

**Comment by marstamm** [comment]
(I renamed the planesBehavior with f38d75e)

**Comment by nikku** [comment]
But what about https://github.com/bpmn-io/diagram-js/pull/601#issuecomment-990849278?

**Review by marstamm — COMMENTED** [review]
I missed that one, sorry! I somehow still had the "we reworked it to work on base layer only" in my head

I have the following proposal:
- we handle everything over the root elements, including the association to the internal plane element
- let's move away from the association of layer name = root element name
- Instead, reuse the Index of the Root element for the layer name

I'll add a proposal commit for your consideration

**Comment by nikku** [comment]
> https://github.com/bpmn-io/diagram-js/pull/601#pullrequestreview-828889102

As long as it works with modeling I#m fine with whatever :100:

**Comment by marstamm** [comment]
See 997951930b28c40ea86ad5d5daa7af5a8f383a2f for modeling capabilities. We don't have to rename layers when we just number them

**Comment by nikku** [comment]
Following up on this, why don't we simply store the `layerName` on the `rootElement`, rather than the `idx`?

**Comment by nikku** [comment]
Cf. https://github.com/bpmn-io/diagram-js/pull/601/commits/36994096bcb6ed299c436284f86655d6b32b67fe.

**Comment by marstamm** [comment]
There was no particular reason for it. Saving the layer name on the element works as well :+1:

**Comment by nikku** [comment]
Cleaned up and marked as review once more.

**Review by marstamm — APPROVED** [review]

**Pull request merged**
